query_executor.go

```diff
@@ -108,96 +108,106 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
 }
 
 func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter NextHost) *Iter {
-	selectedHost := hostIter()
 	rt := qry.retryPolicy()
-	lwt_rt, use_lwt_rt := rt.(LWTRetryPolicy)
-	// We only want to apply LWT policy to LWT queries
-	use_lwt_rt = use_lwt_rt && qry.IsLWT()
+	lwtRT, isLWTRT := rt.(LWTRetryPolicy)
+
+	var getShouldRetry func(qry RetryableQuery) bool
+	var getRetryType func(error) RetryType
+	if rt == nil {
+		getShouldRetry = func(qry RetryableQuery) bool {
+			return false
+		}
+		getRetryType = func(err error) RetryType {
+			return RetryNextHost
+		}
+	} else if isLWTRT && qry.IsLWT() {
+		getShouldRetry = lwtRT.AttemptLWT
+		getRetryType = lwtRT.GetRetryTypeLWT
+	} else {
+		getShouldRetry = rt.Attempt
+		getRetryType = rt.GetRetryType
+	}
 
-	var lastErr error
-	var iter *Iter
-	var conn *Conn
 	var potentiallyExecuted bool
-	for selectedHost != nil {
+
+	execute := func(qry ExecutableQuery, selectedHost SelectedHost) (iter *Iter, retry bool) {
 		host := selectedHost.Info()
 		if host == nil || !host.IsUp() {
-			if rt == nil {
-				selectedHost = hostIter()
-				lastErr = ErrHostDown
-				continue
-			}
-		} else {
-			pool, ok := q.pool.getPool(host)
-			if !ok {
-				if rt == nil {
-					selectedHost = hostIter()
-					lastErr = ErrNoPool
-					continue
-				}
-			} else {
-				conn = pool.Pick(selectedHost.Token(), qry)
-				if conn == nil {
-					if rt == nil {
-						selectedHost = hostIter()
-						lastErr = ErrNoConnectionsInPool
-						continue
-					}
-				} else {
-					iter = q.attemptQuery(ctx, qry, conn)
-					iter.host = selectedHost.Info()
-					// Update host
-					switch {
-					case errors.Is(iter.err, context.Canceled),
-						errors.Is(iter.err, context.DeadlineExceeded),
-						errors.Is(iter.err, ErrNotFound):
-						// those errors represents logical errors, they should not count
-						// toward removing a node from the pool
-						selectedHost.Mark(nil)
-						if potentiallyExecuted && !qry.IsIdempotent() {
-							iter.err = &QueryError{err: iter.err, potentiallyExecuted: true, isIdempotent: false}
-						}
-						return iter
-					default:
-						selectedHost.Mark(iter.err)
-					}
-
-					// Exit if the query was successful
-					// or no retry policy defined
-					if iter.err == nil || rt == nil {
-						return iter
-					}
-
-					lastErr = iter.err
-
-					if customErr, ok := iter.err.(*QueryError); ok && customErr.PotentiallyExecuted() {
-						customErr.isIdempotent = qry.IsIdempotent()
-						lastErr = customErr
-						potentiallyExecuted = true
-					}
-				}
-			}
+			return &Iter{
+				err: &QueryError{
+					err:                 ErrHostDown,
+					potentiallyExecuted: potentiallyExecuted,
+				},
+			}, getShouldRetry(qry)
+		}
```

Comment:
with this change in this case if rt = nil we would not retry, but it is not in line with current behavior, when we retry on the next host, I would change `getShouldRetry(qry)` to just `true` in first three cases